content/blog/angular-components-control-value-accessor/index.md

```diff
@@ -0,0 +1,419 @@
+
+Something you'll notice that wasn't implemented in the `ControlValueAccessor` implementation is support for checking whether validators are applied. If you're a well-versed Angular Form-ite, you'll recall the ability to [validate forms using validators appended to `FormControl`s](https://angular.io/guide/form-validation). Although a niche situation — since most validation happens at the page level, not the component level — wouldn't it be nice to check when a form is valid or not directly from the component to which the form is attached?
+
+Well, thanks to Angular's DI system, we can do just that!
+
+However, we'll need to make a few changes to the form input [that we made before](#forwardRef). While we previously implemented a provider for form controls, we now need to manually assign the provider ourselves in the constructor:
+
+```typescript
+import {
+  Component,
+  Input,
+  ChangeDetectorRef,
+  Optional,
+  Self,
+  AfterContentInit
+} from "@angular/core";
+import { ControlValueAccessor, NgControl } from "@angular/forms";
+
+@Component({
+  selector: "app-example-input",
+  templateUrl: "./example-input.component.html",
+  styleUrls: ["./example-input.component.css"]
+})
+export class ExampleInputComponent implements ControlValueAccessor, AfterContentInit {
+  constructor(
+    @Optional() @Self() public ngControl: NgControl,
+    private _changeDetector: ChangeDetectorRef
+  ) {
+    if (ngControl != null) {
+      // Setting the value accessor directly (instead of using
+      // the providers) to avoid running into a circular import.
+      ngControl.valueAccessor = this;
+    }
+  }
+    
+  // ...
+}
+```
+
+In this code sample, we're using [the `@Self` decorator](https://angular.io/api/core/Self) to tell the dependency injection system that "this component _itself_ should have been provided a `formControl` or `formControlName`". However, we want the component to work even when `FormModule` isn't being used, so we allow the dependency injection to return `null` if nothing's passed by utilizing [the `@Optional` decorator](https://angular.io/api/core/Optional).
+
+Now that you have the `ngControl`, you can access the `formControl` by using `ngControl.control`.
+
+```typescript
+ngOnInit() {
+    const control = this.ngControl && this.ngControl.control;
+    if (control) {
+        console.log("ngOnInit", control);
+        // FormControl should be available here
+    }
+}
+```
+
+You have [a ton of different props you're able to access for the control's metadata](https://angular.io/api/forms/NgControl). For example, if you want to check when errors are present, you can do the following:
+
+```typescript
+get errors() {
+	const control = this.ngControl && this.ngControl.control;
+    if (control) {
+    	return control.touched && control.errors;
+    }
+    return null;
+}
+```
+
+And then reference it in the template:
+
+```html
+<span class="inputLabel" [class.redtext]="errors">{{ placeholder }}</span>
```

Comment:
While you can do multiple classes this way, you could also do it with templating
```suggestion
<span class="inputLabel{{errors ? ' redtext' : ''}}">{{ placeholder }}</span>
```

This one is almost entirely personal taste though. I like this method better but a lot of people don't because of the empty string at the end

Reply:
I'd rather keep things the way we are